README.md

# Juniper Networks vMX lwaftr Docker Container

The vmx-docker-lwaftr Docker Container contains everything thats required to successfully launch vMX 16.1R3 and newer images with a configuration file and license key. This document describes how that Container can be built from source. The actual vMX images is NOT part of the Container. It will be loaded from the official vMX tar file placed in the local directory from where the Container is launched.

![Architecture](vmx-docker-light.png)

Consult the Juniper White Paper on [vMX Lightweight 4over6 Virtual Network Function](https://www.juniper.net/assets/us/en/local/pdf/whitepapers/2000648-en.pdf) for a solution overview and listen to the podcast on Software Gone Wild by Ivan Pepelnjak, Dec 2016: [Blog](http://blog.ipspace.net/2016/12/snabb-switch-with-vmx-control-plane-on.html), [MP3](http://stream.ipspace.net/nuggets/podcast/Show_68-lwAFTR_Snabb_Data_Plane_with_vMX_Control_Plane.mp3).

## Requirements

- [Juniper vMX 17.3 or newer](http://www.juniper.net/support/downloads/?p=vmx)
- Linux kernel 3.13.0+, e.g. as part of Ubuntu 14.04+ or any other Linux distribution
- [Docker Engine 1.12.0+](https://docs.docker.com/engine/installation/)
- [Docker Compose 1.14.0](https://docs.docker.com/compose/install/)
- make (install this e.g. via sudo apt-get make)
- Hugepages activated (2M or 1GB size are fine)
- Kernel modules loaded: kvm, loop

## Build instructions

The Container vmx-docker-lwaftr is based on the official Ubuntu Docker 14.04 base Container and includes the following elements:

* Qemu 
* Snabb
* JET Python Client Library

Starting with version 1.2.0, the build process has been simplified, thanks to docker-compose and
Junos 17.3, which allows the forwarding engine to run natively in the Linux container and hence
removes the need for a custom Qemu version. Instead of submodules, specific branches are pulled 
and built as part of the build process.

### Clone repo

[wip] until a 1.2.0 tag is set, clone branch criot from mwiget instead of Juniper.

```
git clone -b criot https://github.com/mwiget/vmx-docker-lwaftr
```

Or to clone a specific version by tag, use git option '-b' to specify a specific tag:

```
git clone -b v1.2.0 https://github.com/juniper/vmx-docker-lwaftr
```

### Install docker-compose

Check if your compute node has docker-compose 1.14.0 or newer installed:

```
$ docker-compose --version
docker-compose version 1.14.0, build c7bdf9e
```

Otherwise download it:

```
$ sudo su -
# curl -L https://github.com/docker/compose/releases/download/1.14.0/docker-compose-`uname -s`-`uname -m` > /usr/local/bin/docker-compose
# chmod a+rx /usr/local/bin/docker-compose
# exit
$ docker-compose --version
docker-compose version 1.14.0, build c7bdf9e
```

### Build Containers

The build process is now managed via the [docker-compose.yml](docker-compose.yml) file. 
In addition, a [Makefile](Makefile) is also provided to drive the various docker-compose tools
with make, but its use is optional.

```
docker-compose build
```

This will take several minutes and requires Internet access. Alternatively, use 'make build'.
Use 'docker-compose images' to 
show the various containers built:


```
$ docker-compose images
Container                       Repository              Tag       Image Id      Size
-------------------------------------------------------------------------------------------------
vmxdockerlwaftr_b4cpe_1           vmxdockerlwaftr_b4cpe           latest   62dcf1646dd4   248 MB
vmxdockerlwaftr_lwaftr_1          vmxdockerlwaftr_lwaftr          latest   93ced775afe5   510 MB
vmxdockerlwaftr_packetblaster_1   vmxdockerlwaftr_packetblaster   latest   efc7db6754eb   13.7 MB
vmxdockerlwaftr_server_1          vmxdockerlwaftr_server          latest   6c60bb084491   18.5 MB
```

| Container Image | Description |
|---|--|
| vmxdockerlwaftr_lwaftr_1 | Main container to launch Junos vMX image |
| vmxdockerlwaftr_packetblaster_1 | Generates low PPS traffic into xe-0/0/0 over linux bridges |
| vmxdockerlwaftr_b4cpe_1 | B4 test client for one subscriber connected to xe-0/0/1 |
| vmxdockerlwaftr_server_1 | Test server for the B4 client to ping and query via http at 1.1.1.1 |

### Save vmx-docker-lwaftr Container to file

Optional. Save the vmx-docker-lwaftr Container into an image file for transfer to a compute node without Internet connectivity:

```
$ docker save -o vmx-docker-lwaftr-v1.2.0.img vmxdockerlwaftr_lwaftr
mwiget@sa:~/vmx-docker-lwaftr$ ls -l vmx-docker-lwaftr-v1.2.0.img
-rw------- 1 mwiget mwiget 541005824 Jul 16 15:50 vmx-docker-lwaftr-v1.2.0.img
```

### Install hugepages

Edit file /etc/default/grub and update the kernel option line:

```
GRUB_CMDLINE_LINUX_DEFAULT="nomodeset hugepages=6000 intel_iommu=off"
$ sudo update-grub
$ sudo reboot
```

Default hugepage size is 2M, so 6000 reserves 12GB of RAM. Make sure the reservation doesn't exceed the system's memory. Then reboot the system.

Check if the hugepages are allocated properly:

```
$ cat /proc/meminfo |grep Huge
AnonHugePages:    903168 kB
ShmemHugePages:        0 kB
HugePages_Total:    6000
HugePages_Free:     5982
HugePages_Rsvd:        0
HugePages_Surp:        0
Hugepagesize:       2048 kB
```

### Kernel modules

Check if the loop mount kernel module is loaded and load it otherwise:
```
$ lsmod|grep loop
mwiget@sb:~$ sudo modprobe loop
mwiget@sb:~$ lsmod|grep loop
loop                   28672  0
```

Do the same for the kvm_intel kernel module:

```
$ lsmod|grep kvm
kvm_intel             192512  3
kvm                   589824  1 kvm_intel
irqbypass              16384  1 kvm
```

To make both changes permanent, add the modules to /etc/modules:

```
$ cat /etc/modules
# /etc/modules: kernel modules to load at boot time.
#
# This file contains the names of kernel modules that should be loaded
# at boot time, one per line. Lines beginning with "#" are ignored.
loop
kvm_intel
```

## Running the vmx-docker-lwaftr Container

Before launching it, you need provide the Junos qcow2 image from the vmx-bundle tar file, a license key and a ssh public key to access the instance once running.
Place these three files into the top directory of the cloned repo.

### Download and extract junos*qcow2

Download vMX 17.3 or newer from Juniper, then extract the junos*qcow2 Junos image. Nothing else is required. The forwarding engine will be downloaded from the control plane automatically. 

```
$ tar zxf vmx-bundle-17.3B1.1.tgz
$ mv vmx/images/junos-vmx-x86-64-17.3B1.1.qcow2 .
$ rm -rf vmx
$ ls -l junos-vmx-x86-64-17.3B1.1.qcow2
-rw-r--r-- 1 mwiget mwiget 1226440704 Jun 22 00:14 junos-vmx-x86-64-17.3B1.1.qcow2
```

### Download license key

Download the [vMX eval license key from Juniper](https://www.juniper.net/us/en/dm/free-vmx-trial/):

```
$ wget  https://www.juniper.net/us/en/dm/free-vmx-trial/E421992502.txt
--2017-07-16 16:05:40--  https://www.juniper.net/us/en/dm/free-vmx-trial/E421992502.txt
Resolving www.juniper.net (www.juniper.net)... 2a02:26f0:78:198::720, 2a02:26f0:78:181::720, 95.101.252.154
Connecting to www.juniper.net (www.juniper.net)|2a02:26f0:78:198::720|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 206 [text/plain]
Saving to: ‘E421992502.txt’

E421992502.txt            100%[==================================>]     206  --.-KB/s    in 0s

2017-07-16 16:05:40 (359 MB/s) - ‘E421992502.txt’ saved [206/206]

$ mv E421992502.txt license-eval.txt
```

If you have a production key, use that one instead and adjust the license key file name in the [docker-compose.yml](docker-compose.yml) file.

### Copy your ssh public key

Each instance will create a random root password, accessible via 'docker logs <instance-id>'. To access the instance via ssh public/private key, copy your ssh public key file to the repo root directory:

```
$ cp ~/.ssh/id_rsa.pub .
```

### Launch instance

The launch of the instance is done via 'docker-compose up -d'. The option '-d' is needed to launch the instance in the background. The use of 'make up' is a shorter version to type, but uses the same command. It builds the container image again if needed too.

Check the content of [docker-compose.yml](docker-compose.yml) to understand what exactly gets built and launched. It serves as an example and will need adjustment. "out-of-the-box" it launches one lwaftr (vMX) instance, a packetblaster traffic generator for 60k subscribers into 0/0/0 and a b4cpe client, which can be used to interactively test the functionality via 0/0/1 by pinging and connecting to 1.1.1.1 via HTTP.

```
$ make up
docker-compose build
Building lwaftr
Step 1/22 : FROM ubuntu:14.04
 ---> 4a2820e686c4
Step 2/22 : LABEL maintainer Juniper Networks
 ---> Using cache
 ---> b304d31999c8
. . .
Successfully built d3b692e019aa
Successfully tagged vmxdockerlwaftr_server:latest
docker-compose up -d
Creating network "vmxdockerlwaftr_default" with the default driver
Creating network "vmxdockerlwaftr_net-xe0" with the default driver
Creating network "vmxdockerlwaftr_net-xe1" with the default driver
Creating vmxdockerlwaftr_lwaftr_1 ...
Creating vmxdockerlwaftr_lwaftr_1 ... done
Creating vmxdockerlwaftr_packetblaster_1 ...
Creating vmxdockerlwaftr_server_1 ...
Creating vmxdockerlwaftr_b4cpe_1 ...
Creating vmxdockerlwaftr_b4cpe_1
Creating vmxdockerlwaftr_server_1
Creating vmxdockerlwaftr_server_1 ... done
$
```

This creates the required management network, networks with default driver for xe0 and xe1, a packetblaster, b4cpe client and server to test the functionality, and finally the vMX itself. 

The [Makefile](Makefile) contains shortcuts to various useful functions. It is worth to explore this file to understand how these commands can be used directly.

### Check the status 

```
$ make ps
docker-compose ps
             Name                            Command               State                       Ports
------------------------------------------------------------------------------------------------------------------------
vmxdockerlwaftr_b4cpe_1           /launch.sh 193.5.1.2 1024- ...   Up
vmxdockerlwaftr_lwaftr_1          /usr/bin/dumb-init -- /usr ...   Up      0.0.0.0:32905->22/tcp, 0.0.0.0:32904->830/tcp
vmxdockerlwaftr_packetblaster_1   /launch.sh --b4 2001:db8:: ...   Up
vmxdockerlwaftr_server_1          /launch.sh                       Up
```

If one of the instances is down, check its log file. For the main vmxdockerlwaftr_lwaftr_1 instance, a short cut is provided:

```
$ make logs
docker logs -f $(docker ps -a|grep _lwaftr|cut -d' ' -f1)
Juniper Networks vMX lwaftr Docker Container Sun Jul 16 14:12:59 UTC 2017

/u contains the following files:
Changelog.md	 b4cpe			id_rsa.pub			 server
LICENSE		 binding_table_60k.txt	junos-vmx-x86-64-17.3B1.1.qcow2  snabb
Makefile	 docker-compose.yml	license-eval.txt		 src
README.md	 dumb-init		lwaftr.conf.txt
SUPPORT-INFO.md  getpass.sh		packetblaster
Launching with arguments:
02:42:ac:14:01:10 02:42:ac:14:02:10
02:42:ac:14:01:10 eth0 -> eth1
02:42:ac:14:02:10 eth0 -> eth2
-----------------------------------------------------------------------
vMX vmxdockerlwaftr_lwaftr_1 (172.18.0.2) root password laenohneiraiquiucujehuro
-----------------------------------------------------------------------

using qcow2 image junos-vmx-x86-64-17.3B1.1.qcow2
using license keys in license-eval.txt
ethernet interfaces: eth0 eth1 eth2
creating ssh public/private keypair to communicate with Junos
Generating public/private rsa key pair.
. . .
```


Changelog.md
## v1.2.0:

- Native forwarding engine (riot). Requires Junos 17.3
- Replaced submodules with build stages in Dockerfile (no more git --recursive)
- Using docker-compose to build and run
- No longer custom compile qemu, using ubuntu package (it was required for VFP, which has 
  been replaced with riot running native in the container
- Specify junos-vmx-x86-64-17.3B1.1.qcow2 image instead of vmx-bundle for faster startup
- No longer requires identity private ssh key: auto-generates unique ssh pair at startup
  to communicate between the JET apps and Junos
- auto-provision interface IP for virtual interfaces and loopback 

## v1.1.21:

- Support for 16.1R4.7 and 17.1R1.8 added (by limiting JSON fix to 16.1R3)
- Enable lite PFE mode in tests/lwaftr5.txt

## v1.1.20:

- Add --mirror flag to snabbvmx (use 'snabb lwaftr monitor' to activate or disable)
- Removed monitor.sh script. 
- Added lite-mode to lwaftr1.txt configuration

## v1.1.19:

- Use latest Ubuntu base container

## v1.1.18:

- Move to using ubuntu 14.04 base container instead of specific minor version
  due to build conflicts with outdated containers

## v1.1.18:

- update snabb with binding table reload fix

## v1.1.17:

- updated test scripts to 16.1R3.10
- Add white paper link to README.md
- Added copyright and breaths count to the Snabb show and monitor CLIs
- Added MIB support for breath counters
- Added logs for the snmp script
- Fixed the breath display issue
- fix parsing of binding table filename

## v1.1.16:

- Update snabb to 9c07396: Let out-ipv6-* counters also count internally generated ICMPv6 packets, and update docs (#547))
- remove stats.xml from collector. No longer exists

## v1.1.15:

- Update snabb to bd3811f Add unit test for ipv4_cache_trigger

## v1.1.14:

- set icmpv6_rate_limiter_n_packets and icmpv6_rate_limiter_n_seconds in .conf file
- fixed ingress_drop_action to become ingress_drop_monitor in snabbvmx cfg file

## v1.1.13:

- Fixed typo in the index number for ps counters for monitor lwaftr
- Fixed the negative counter and the merge issue in the igalia branch due to my commit.

## v1.1.12:

- JET Python client package added into jet-python

## v1.1.11:

- Snabb Misc. cleanup of the program/lwaftr/ subtree (#501)
- Snabbb Merge pull request #500 from Igalia/merge-next-to-lwaftr
- Removing the stale print statements for show lwaftr |display xml

## v1.1.10:

- Fixed xml output for 'show lwaftr statistics | disolay xml'

## v1.1.9:

- Updated snabb submodule to commit 'Remove old benchmarking tools'
- Updated snabb submodule with fix for 'Snabbvmx ignores vlan settings in .conf file'
- Updated snabb submodule to [Igalia/snabb] Merge v2016.10 (#483)
- lwaftr5 test case to use vlan tag on 0/0/1

## v1.1.8:

- set vlan tag info in lwaftr .conf file, not snabbvmx .cfg
- fixed Error: File "/tmp/snabb_xe0.log" is empty.

## v1.1.7:

- check if vMX has been given any interfaces and terminate if not

## v1.1.6:

- Sync Snabb to igalia/lwaftr commit f6272b03864705dc401f088d6d87c4e3ed2fba0b, Oct 10
- Add fragmentation config knobs in YANG for lwaftr app
- fix typo (Statisitics)

## v1.1.5:

- Increase vMX RE /config disk to 6GB (from 2GB)

## v1.1.4:

- snabb syncd to igalia/lwaftr 7277beca69e50536a1a6d8e861301855e5220a40 (Fri Oct 7 11:42:37 2016 +0200)
- YANG package ID limited to 5 characters (set to lwaft from lwaftr)
- fix monitor lwaftr v4 per second counters
- document lab and root password in test configuration files (lab/lab123 root/lab123)
- Fixed minor issue in the lwaftr monitor & a mutex issue in the daemonized app.
- use snabb_stats.xml provided by JET
- changed formatting output in monitor lwaftr
- use sudo in test scripts for snabb
- fix dynamic rendering for counters in show lwaftr statistics
- enhance support-info doc
- let JET daemon do the statistics gathering
- create test scripts for offline support use

## v1.1.3:

- Updated scripts and yang files for JET

## v1.1.2:

- add submodules and combine bulid in one container 
- remove nexthop.sh. Use snabbvmx query instead
- remove reference to removed nexthop.sh
- updated JET op and snmp script
- remove obsolete directory python-tools
- update Dockerfile with correct snmp python script
- Merge pull request #459 from Igalia/refactor-lwutils
- updated jet scripts
- snabb: latest commit to include Merge pull request #471 from mwiget/snabbvmx-query-id-name
- update Changelog with v1.1.2

## v1.1.0:

- fix lwaftr4 config and use of daily build for testing
- add copyright notice
- add JET opserver and show scripts
- upgrade snabb to snabbvmx-v1.1, tag as v1.1.0

## v1.0.2:

- collect snabb binary with collect-support-infos.sh
- update SUPPORT-INFO.md with adding snabb binary to tgz file
- make jnx-aug-softwire:ipv4_address mandatory
- fix detecting binding table file changes in snabbvmx_manager.pl
- upgraded configs to 16.1R2

## v1.0.1:

- fix numanode warning when launched with 82599 NICs on socket > 0
- Merge branch 'igalia' of https://github.com/mwiget/vmxlwaftr into igalia
- to include snabb Fix update of ingress-packet-drops counter #449
- fix condition when DEBUG isn't defined
- collect also /tmp/config.* files with collect-support-infos.sh
- create empty binding table if instance doesn't have any defined
- remove snabbvmx configs when br instance is no longer configured
- added description to various augmented softwire leafs

## v1.0:
- using Snabb tag lwaftr-snabbvmx-v1.0 from https://github.com/igalia/snabb


src/launch.sh
#!/bin/bash
# Copyright (c) 2017, Juniper Networks, Inc.
# All rights reserved.
#


VCPMEM="${VCPMEM:-1024}"
VCPU="${VCPU:-1}"
NUMANODE="${NUMANOdE:-0}"

echo -n "Juniper Networks vMX lwaftr Docker Container "
cat /VERSION
echo ""

set -e	#  Exit immediately if a command exits with a non-zero status.

echo "/u contains the following files:"
ls /u

echo "Launching with arguments: $@"

while getopts "h?V:m:c:l:p:R:d" opt; do
  case "$opt" in
    h|\?)
      show_help
      exit 1
      ;;
    V)  VCPU=$OPTARG
      ;;
    m)  VCPMEM=$OPTARG
      ;;
    c)  CONFIG=$OPTARG
      ;;
    l)  LICENSE=$OPTARG
      ;;
    p)  PUBLICKEY=$OPTARG
      ;;
    R)  QEMUVCPCPUS=$OPTARG
      ;;
    d)  DEBUG=1
      ;;
  esac
done

CONFIG="${CONFIG-config.txt}"

shift "$((OPTIND-1))"

if [ ! -z "$1" ]; then
  IMAGE=$1
  shift
fi

# env beats command line arguments:
PCI_INTERFACES="${INTERFACES:-$@}"

if [ ! -f "/u/$IMAGE" ]; then
  echo "vMX tar file $IMAGE not found"
  exit 1
fi

HPFREE=$(cat /proc/meminfo |grep HugePages_Free|awk {'print $2'})
if [ $HPFREE = 0 ]; then
   echo 'no free HugePages found. Need those to run forwarding engine:'
   cat /proc/meminfo|grep Huge
   exit 1
fi

# fix network interface order due to https://github.com/docker/compose/issues/4645
/fix_network_order.sh

ROOTPASSWORD=$(pwgen 24 1)
SALT=$(pwgen 8 1)
HASH=$(openssl passwd -1 -salt $SALT $ROOTPASSWORD)
myip=$(ifconfig eth0|grep 'inet addr'|cut -d: -f2|awk '{print $1}')
# extract container name via id
hostname=$(docker ps --format '{{.Names}}' -f id=$HOSTNAME)
hostname="${hostname:-$HOSTNAME}" # in case the name extract magic failed
id=$(echo $hostname|rev | cut -d'_' -f 1 | rev)   # get index, e.g. 3 from vmxdockerlight_vmx_3
export myip hostname id
echo "-----------------------------------------------------------------------"
echo "vMX $hostname ($myip) root password $ROOTPASSWORD"
echo "-----------------------------------------------------------------------"
echo ""


if [[ "$IMAGE" =~ \.qcow2$ ]]; then
  echo "using qcow2 image $IMAGE"
  cp /u/$IMAGE /tmp/
  VCPIMAGE=$IMAGE
else
  echo "extracting qcow2 image from $IMAGE ..."
  tar zxvf /u/$IMAGE -C /tmp/ --wildcards vmx/images/junos*qcow2 2>/dev/null
  VCPIMAGE=$(ls /tmp/vmx*/images/junos*qcow2)
  mv $VCPIMAGE /tmp/
  VCPIMAGE=${VCPIMAGE##*/}
fi

if [ ! -f "/u/$LICENSE" ]; then
  echo "Warning: No license file found ($LICENSE)"
else
  echo "using license keys in $LICENSE"
fi

# loop thru ethernet interfaces and bridge eth0 to tap fxp0, remove its IP
# and build matching junos config

ethlist=$(netstat -i|grep eth|cut -d' ' -f1 | paste -sd " " -)
echo "ethernet interfaces: $ethlist"

echo "creating ssh public/private keypair to communicate with Junos"
ssh-keygen -t rsa -f /root/.ssh/id_rsa -N ''
JETPUBKEY=$(cat /root/.ssh/id_rsa.pub)
IDENTITY="/root/.ssh/id_rsa"

if [ ! -f "/u/$PUBLICKEY" ]; then
  echo "WARNING: Can't read ssh public key file $PUBLICKEY"
  SSHPUBLIC=$JETPUBKEY  # to keep junos config parser happy
else
  SSHPUBLIC=$(cat /u/$PUBLICKEY)
fi

cat > /tmp/$CONFIG <<EOF
system {
  host-name $hostname;
  root-authentication {
    encrypted-password "$HASH";
    ssh-rsa "$SSHPUBLIC";
    ssh-rsa "$JETPUBKEY";
  }
  services {
    ssh {
      client-alive-interval 30;
    }
    netconf {
      ssh;
    }
  }
  syslog {
    file messages {
      any notice;
    }
  }
  extensions {
    extension-service {
      application {
        file rpc-jet.py {
          daemonize;
          username root;
        }
      }
    }
  }
}
interfaces {
lo0 {
  unit 0 {
    family inet {
      address 127.0.0.1/24;
    }
  }
}
}
EOF

# auto-add op scripts to config
opfiles=$(ls /op/*slax)
if [ ! -z "$opfiles" ]; then
  cat >> /tmp/$CONFIG <<EOF
system {
  scripts {
    op {
EOF
  for file in $opfiles; do
    file=${file##*/}
    echo "adding op script $file to config"
    echo "      file $file;" >> /tmp/$CONFIG
  done
  cat >> /tmp/$CONFIG <<EOF
    }
    language python;
  }
}
EOF
fi

if [ ! -z "$SSHPUBLIC" ]; then
  SSHUSER=$(echo $SSHPUBLIC | cut -d' ' -f3 | cut -d'@' -f1)
  echo "adding super-user $SSHUSER with public key $SSHPUBLIC to config"
  cat >> /tmp/$CONFIG <<EOF
system {
  root-authentication {
    ssh-rsa "$SSHPUBLIC";
  }
EOF
  if [ "$SSHUSER" != "root" ]; then
    # only do this for non-root logins
    cat >> /tmp/$CONFIG <<EOF
  login {
    user $SSHUSER {
      class super-user;
      authentication {
        encrypted-password "$HASH";
        ssh-rsa "$SSHPUBLIC";
      }
    }
  }
EOF
  fi
  cat >> /tmp/$CONFIG <<EOF
}
EOF
fi
cat >> /tmp/$CONFIG <<EOF
groups {
  docker-networking {
    protocols {
      lldp {
        interface all;
      }
    }
    interfaces {
EOF

ifdescrlist=$(docker inspect --format='{{range $p, $conf := .NetworkSettings.Networks}} {{$p}}  {{end}}' $HOSTNAME)
echo "ifdescr array = $ifdescrlist"
IFS=' ' read -r -a ifdarray <<< "$ifdescrlist"

mygw=$(ip -4 route list 0/0 |cut -d' ' -f3)

index=0
ifdindex=0
for eth in $ethlist; do
  echo "$eth ..."
  myip=$(ifconfig $eth|grep 'inet addr'|cut -d: -f2|awk '{print $1}')
  myipmask=$(ip -o -f inet addr show $eth |awk '{print $4}')
  ifdescr="${ifdarray[$ifdindex]}"
  ifdindex=$(($ifdindex + 1))
  echo "ifdescr=$ifdescr"
  if [ "eth0" == $eth ]; then
    mymac=$(ifconfig $eth |grep HWaddr|awk {'print $5'})
    echo "Bridging $eth ($myipmask/$mymac) with fxp0 (mygw $mygw)"
    brctl addbr br-ext
    ip link set up br-ext
    ip tuntap add dev fxp0 mode tap
    ifconfig fxp0 up promisc
    macchanger -A eth0
    brctl addif br-ext $eth
    brctl addif br-ext fxp0  
    echo "$eth -> fxp0"
    cat >> /tmp/$CONFIG <<EOF
      fxp0 {
        unit 0 {
          description "$eth-$ifdescr"
          family inet {
            address $myipmask;
          }
        }
      }
EOF
  else
    echo "$eth -> ge-0/0/$index"
    cat >> /tmp/$CONFIG <<EOF
      ge-0/0/$index {
        unit 0 {
          description "$eth-$ifdescr"
          family inet {
            address $myipmask;
          }
        }
      }
EOF
    index=$(($index + 1))
  fi
  ip addr flush dev $eth
done

if [ ! -z "$mygw" ]; then
  cat >> /tmp/$CONFIG <<EOF
    }
    routing-options {
     static {
       route 0.0.0.0/0 next-hop $mygw;
     }
   }
EOF
else
  cat >> /tmp/$CONFIG <<EOF
   }
EOF
fi

cat >> /tmp/$CONFIG <<EOF
 }
}
apply-groups docker-networking;
EOF

# find numanode to use based on PCI list.
# It will simply use the numanode of the last PCI.
# Using cards on different Nodes is not recommended 

echo "walking PCI interface list: $PCI_INTERFACES"

for DEV in $PCI_INTERFACES; do # ============= loop thru interfaces start
  PCI=${DEV%/*} 
  if [ "eth" != "${PCI:0:3}" ]; then
    CPU=$(cat /sys/class/pci_bus/${PCI%:*}/cpulistaffinity | cut -d'-' -f1 | cut -d',' -f1)
    NODE=$(numactl -H | grep "cpus: $CPU" | cut -d " " -f 2)
    if [ -z "$NUMAPREV" ]; then
      NUMAFIRST=$NODE
    fi
    if [ "$NODE" != "$NUMAFIRST" ]; then 
      echo "WARNING: Interface $PCI is on numa node $NODE, but first PCI interface is on node $NUMAFIRST"
    else
      echo "Interface $PCI is on node $NODE"
    fi
  fi
done
if [ ! -z "$NUMAFIRST" ]; then
  NUMANODE=$NUMAFIRST
fi

# append user provided config. Doing this after our initial settings ubove
# allows a user to overwrite our defaults, like host-name
if [ -f /u/$CONFIG ]; then
    cat /u/$CONFIG >> /tmp/$CONFIG
fi

mkdir /var/run/snabb
numactl --membind=$NUMANODE mount -t tmpfs -o rw,nosuid,nodev,noexec,relatime,size=4M tmpfs /var/run/snabb

brctl addbr br-int
ip addr add 128.0.0.16/8 dev br-int
ip link set up br-int
ip tuntap add dev em1 mode tap
ifconfig em1 up promisc
brctl addif br-int em1
brctl show

CFGDRIVE=/tmp/configdrive.qcow2
echo "Creating config drive $CFGDRIVE"
/create_config_drive.sh $CFGDRIVE /tmp/$CONFIG /u/$LICENSE 

/usr/sbin/rsyslogd

HDDIMAGE="/tmp/vmxhdd.img"
echo "Creating empty $HDDIMAGE for VCP ..."
qemu-img create -f qcow2 $HDDIMAGE 7G >/dev/null

echo "Building virtual interfaces and bridges for $@ ..."

INTNR=0	# added to each tap interface to make them unique
INTID="xe"

MACP=$(printf "02:%02X:%02X:%02X:%02X" $[RANDOM%256] $[RANDOM%256] $[RANDOM%256] $[RANDOM%256])

CPULIST=""  # collect cores given to PCIDEVS
ETHLIST=$(ifconfig|grep ^eth|grep -v eth0|cut -f1 -d' '|tr '\n' ' ')
LIST="$ETHLIST $PCI_INTERFACES"
echo "walking thru list of interfaces: $PCI_INTERFACES $ETHLIST ..."
for DEV in $LIST; do # ============= loop thru interfaces start

  # 0000:05:00.0/7 -> PCI=0000:05:00.0, CORE=7
  CORE=${DEV#*/}
  PCI=${DEV%/*} 
  PCI=${PCI//0000:/}    # get rid of leading 0000:

  INT="${INTID}${INTNR}"

  # create persistent mac address based on host-name in junos config file
  h=$(grep host-name /tmp/$CONFIG |md5sum)
  if [ "eth" == "${PCI:0:3}" ]; then
    macaddr="02:${h:0:2}:${h:2:2}:${h:4:2}:00:0$INTNR"
    CORE=""
    ifconfig $PCI mtu 9500
  else
    macaddr="02:${h:0:2}:${h:2:2}:${h:4:2}:${PCI:0:2}:0${PCI:6:1}"
    if [ "$CORE" -ge "0" ]; then
      echo "CORE=($CORE) PCI=($PCI)"
    else
      echo "FATAL configuration errror. Must specify core after PCI: $PCI/<core>"
      exit 1
    fi
    if [ -z "$CPULIST" ]; then
      CPULIST=$CORE
    else
      CPULIST="$CPULIST,$CORE"
    fi
  fi

  if [ "eth" != "${PCI:0:3}" ]; then
    if [ -z "$(lspci | grep 82599 | grep $PCI)" ]; then
      echo "No PCI hardware found at $PCI, skipping it"
      continue
    fi
  fi
  echo "PCI=$PCI CORE=$CORE CPULIST=$CPULIST"
  # add PCI to list
  PCIDEVS="$PCIDEVS $PCI"
  INTLIST="$INTLIST $INT"


  echo "$PCI/$CORE" > /tmp/pci_$INT
  echo "$macaddr" > /tmp/mac_$INT

  TAP="tap${INTNR}"    # -> tap/monitor interfaces tap0, tap1 etc
  ip tuntap add dev $TAP mode tap
  ifconfig $TAP up
  echo "created tap interface $TAP for monitoring"

  # we run without VM, so the original ethX interface is renamed to ethXin
  # for snabb and we create a new tap interface with the original ethX name for rio
 
  ip link add ext$INT type veth peer name $INT
  ifconfig ext$INT up

  if [ "eth" == "${PCI:0:3}" ]; then
    mymac=$(ifconfig $PCI |grep HWaddr|awk {'print $5'})
    echo "$mymac" > /tmp/mac_$INT
  else
    # for PCI ports we can take the generated one from /tmp/mac_$INT
    mymac=$(cat /tmp/mac_$INT)
  fi
  echo "copy mac address of $PCI to $INT: $mymac"
  ifconfig $INT down
  ifconfig $INT hw ether $mymac up

  INTNR=$(($INTNR + 1))
done # ===================================== loop thru interfaces done
echo "Done walking interface list: PCIDEVS=$PCIDEVS INTLIST=$INTLIST"

if [ -z "$INTLIST" ]; then
  echo "ERROR: no useable interfaces found. INTLIST ($INTLIST) is empty"
  exit 1
fi

QEMUVCPNUMA="numactl --membind=$NUMANODE"
if [ ! -z "$QEMUVCPCPUS" ]; then
  QEMUVCPNUMA="numactl --membind=$NUMANODE --physcpubind=$QEMUVCPCPUS"
fi

# calculate the cpu affinity mask excluding the ones for snabb
AVAIL_CORES=$(taskset -p $$|cut -d: -f2|cut -d' ' -f2)

echo "CPULIST=$CPULIST AVAIL_CORES=$AVAIL_CORES"

if [ ! -z "$CPULIST" ]; then 
  SNABB_AFFINITY=$(taskset -c $CPULIST /usr/bin/env bash -c 'taskset -p $$'|cut -d: -f2|cut -d' ' -f2)
  let AFFINITY_MASK="0x$AVAIL_CORES ^ 0x$SNABB_AFFINITY"
  AFFINITY_MASK=$(printf '%x\n' $AFFINITY_MASK)
  # Note: doesn't work with numactl: it will refuse to use a cpu that is masked out
  #echo "set cpu affinity mask $AFFINITY_MASK for everything but snabb"
  #taskset -p $AFFINITY_MASK $$
  echo "taskset -p $AFFINITY_MASK \$\$" >> /root/.bashrc
fi

BINDINGS=$(grep binding-table-file /tmp/$CONFIG | awk '{print $2}'|cut -d';' -f1)
if [ -f /u/$BINDINGS ]; then
  cp /u/$BINDINGS /tmp/
  BINDINGS=$(basename $BINDINGS)
else
  echo "WARNING: Binding table file $BINDINGS not found"
fi

if [ ! -z "$DEBUG" ]; then
   echo "Debug shell. Exit shell to continue"
   bash
   if [ $? -gt 0 ]; then
     exit
   fi
fi

# Launching snabb processes after we set excluded the cores
# from the scheduler
for INT in $INTLIST; do
  echo "Launching snabb on $INT"
  cd /tmp && numactl --membind=$NUMANODE /launch_snabb.sh $INT &
done

export PFE_SRC
sh /start_pfe.sh &

echo "launching JET opserver ..."
cd /tmp && /launch_opserver.sh &

echo "launching snabbvmx_manager.pl ..."
cd /tmp && /launch_snabbvmx_manager.sh 128.0.0.1 $IDENTITY $BINDINGS &

cd /tmp

qemu-system-x86_64 -M pc --enable-kvm -cpu host -smp $VCPU -m $VCPMEM \
  -smbios type=0,vendor=Juniper \
  -smbios type=1,manufacturer=VMX,product=VM-vcp_vmx1-161-re-0,version=0.1.0 \
  -no-user-config \
  -no-shutdown \
  -drive if=ide,file=$VCPIMAGE -drive if=ide,file=$HDDIMAGE \
  -drive if=ide,file=$CFGDRIVE \
  -device cirrus-vga,id=video0,bus=pci.0,addr=0x2 \
  -netdev tap,id=tc0,ifname=fxp0,script=no,downscript=no \
  -device virtio-net-pci,netdev=tc0,mac=$mymac \
  -netdev tap,id=tc1,ifname=em1,script=no,downscript=no \
  -device virtio-net-pci,netdev=tc1 \
  -nographic || true
